### PR TITLE
add transferownership route

### PR DESCRIPTION
**Pull request by alexcos20**
## Description

Add /api/v1/aquarius/assets/ddo/owner/update/<did> route

### Reason
 Allowing full DDO update for now is risky. There are deep implications (checksum, didregistry, etc)
Instead, allow specific update operations.

## Is this PR related with an open issue?

Related to Issue #

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Important:
***New ENV var: KEEPER_URL***
***New DDO attibute:  updated  - timestamp containing last ddo update***

Signature:
   Message used for signature is the current update field in the DDO.  This assures that the signature cannot be used after, because the updated field is changed after every call
```
@assets.route('/ddo/owner/update/<did>', methods=['PUT'])
parameters:
      - name: did
        in: path
        description: DID of the asset.
        required: true
        type: string
        example: "did:op:d007b84d6f874cbf868177898f2353f7adfc824c9f9843d8b9ee60596db3b9f0"
      - in: body
        name: body
        required: true
        schema:
          type: object
          required:
            - signature
            - updated
            - newOwner
          properties:
            "signature":
              description: Signature using updated field to verify that the consumer has rights to update 
              type: string
              example: "0x42e940108a430b91796341e29001319b2b2c4743156cdbe0e17afdae82b4cf9a7e1b4e641cd57d8f087ab6432cc9e53989f3ce121b6897fa3f594e9753c4ea331b"
            "updated":
              description: Last update field of the DDO
              type: string
              example: "2020-01-01T00:00:00Z"
            "newOwner":
              description: The new owner ethereum address.
              type: string
              example: "0x858048e3Ebdd3754e14F63d1185F8252eF142393"
```

How to sign in nodejs:
```
function gethash(message) {
    var hex = ''
    for(var i=0;i<message.length;i++) {
        hex += ''+message.charCodeAt(i).toString(16)
    }
    var hexMessage = "0x" + hex
    return hexMessage
}

let updated="2020-01-01T00:00:00Z"
let did='did:op:100'
let newowner=0x200

var signature=await web3.eth.personal.sign(gethash(updated), account, null)
```
(explanations for this method of signing: https://ethereum.stackexchange.com/questions/55003/recover-javascript-signed-message-in-python-web3)

This opens that way to have other updates & deletes possible, without having to update the entire DDO, using the following function:
```
_can_update_did(_record,updated,signature)

### CODE REVIEW

**Comment by alexcos20** [comment]
WIP:  have to write tests

File: aquarius/app/assets.py [hunk 1]
```diff
@@ -461,6 +465,84 @@ def update(did):
         return f'Some error: {str(err)}', 500
 
 
+@assets.route('/ddo/transferownership/<did>', methods=['PUT'])
+def transferownership(did):
```
Comment (ssallam) [inline aquarius/app/assets.py]: we have to follow the same function naming convention (use underscores).

File: aquarius/app/assets.py [hunk 2]
```diff
@@ -767,3 +849,49 @@ def _reorder_services(services):
             result.append(service)
 
     return result
+
+def getsigneraddress(message, signature):
```
Comment (ssallam) [inline aquarius/app/assets.py]: same as above, naming convention.

File: README.md [hunk 3]
```diff
@@ -42,6 +42,27 @@ If you want to know more about the ontology of the metadata, you can find all th
...
+```
+Then build the Docker image 
+```bash
+docker build -t "myaqua" .
+```
+Run Docker image
+```bash
+docker run myaqua
+```
+To test with other ocean components in `barge` directory  `compose-files/aquarius.yml`  change `aquatios image` to `myaqua`
```
Comment (ssallam) [inline README.md]: typo in `aquatios image`

File: aquarius/app/assets.py [hunk 4]
```diff
@@ -461,6 +465,84 @@ def update(did):
...
+    data = request.json
+    required_attributes = [
+        'signature',
+        'updated',
+        'newowner'
+    ]
+    msg, status = check_required_attributes(required_attributes, data, 'transferownership')
+    if msg:
+        return msg, status
+    if web3.isAddress(data['newowner'])==False:
```
Comment (ssallam) [inline aquarius/app/assets.py]: if statement can be simplified: `if not web3.isAddress(data['newowner']):`

File: aquarius/app/assets.py [hunk 5]
```diff
@@ -461,6 +465,84 @@ def update(did):
...
+        _record=dao.get(did)
+        if _record is None:
+            return f'Cannot find did: {did} ', 404
+        if not _can_update_did(_record,data['updated'],data['signature']):
+            logger.error('Not allowed to update did')    
+            return f'Not allowed to update this DID', 500
+        if compare_eth_addresses(_record['publicKey'][0]['owner'],data['newowner']):
+            return f'New owner must be different than owner', 500
+        _record['publicKey'][0]['owner']=data['newowner']
+        _record['updated']=get_timestamp()
```
Comment (ssallam) [inline aquarius/app/assets.py]: Use proper spacing.

File: aquarius/app/assets.py [hunk 6]
```diff
@@ -767,3 +849,49 @@ def _reorder_services(services):
...
+        message_hash = defunct_hash_message(text=message)
+        logger.debug('got %s as a message_hash' % message_hash)    
+        address_recovered = web3.eth.account.recoverHash(message_hash, signature=signature)
+        logger.debug('got %s as address_recovered' % address_recovered)    
+        return address_recovered
+    except Exception as e:
+        logger.error(e)
+        return None  
+
+def compare_eth_addresses(address, checker):
```
Comment (ssallam) [inline aquarius/app/assets.py]: This module should only have the endpoints, all other function can be moved to a helper/util module.

File: aquarius/app/assets.py [hunk 7]
```diff
@@ -767,3 +849,49 @@ def _reorder_services(services):
...
+    '''
+    Compare two addresses and return TRUE if there is a match
+    :param str address: Address
+    :param str checker: Address to compare with
+    :return: boolean
+    '''
+    if web3.toChecksumAddress(address) == web3.toChecksumAddress(checker):
```
Comment (ssallam) [inline aquarius/app/assets.py]: simplify: `return web3.toChecksumAddress(address) == web3.toChecksumAddress(checker)`

File: aquarius/app/assets.py [hunk 8]
```diff
@@ -461,6 +465,84 @@ def update(did):
...
+      404:
+        description: Invalid asset data.
+      500:
+        description: Error
+    """
+    data = request.json
+    required_attributes = [
+        'signature',
+        'updated',
+        'newowner'
```
Comment (ssallam) [inline aquarius/app/assets.py]: This name should follow the convension of camel case, newowner -> newOwner

File: aquarius/app/assets.py [hunk 9]
```diff
@@ -461,6 +465,84 @@ def update(did):
         return f'Some error: {str(err)}', 500
 
 
+@assets.route('/ddo/transferownership/<did>', methods=['PUT'])
```
Comment (ssallam) [inline aquarius/app/assets.py]: This is no longer RESTful API.

File: aquarius/app/assets.py (hunk 1 above)
Comment (alexcos20) [reply aquarius/app/assets.py]: fixed

File: aquarius/app/assets.py (hunk 2 above)
Comment (alexcos20) [reply aquarius/app/assets.py]: fixed

File: README.md (hunk 3 above)
Comment (alexcos20) [reply README.md]: fixed

File: aquarius/app/assets.py (hunk 4 above)
Comment (alexcos20) [reply aquarius/app/assets.py]: fixed

File: aquarius/app/assets.py (hunk 5 above)
Comment (alexcos20) [reply aquarius/app/assets.py]: fixed

File: aquarius/app/assets.py (hunk 7 above)
Comment (alexcos20) [reply aquarius/app/assets.py]: fixed

File: aquarius/app/assets.py (hunk 8 above)
Comment (alexcos20) [reply aquarius/app/assets.py]: fixed

File: aquarius/app/assets.py (hunk 9 above)
Comment (alexcos20) [reply aquarius/app/assets.py]: fixed

File: aquarius/app/assets.py (hunk 6 above)
Comment (alexcos20) [reply aquarius/app/assets.py]: fixed

**Comment by alexcos20** [comment]
1. Added Update ratings for a DID
2. Allow DELETE of a DID, providing a signature
3. Allow Metadata Update from marketplace (title, description,sample,prices)
4. Allow whitelists (accessWhiteList and freeWhiteList)

**Comment by alexcos20** [comment]
tested with rating_backend, freezing for now.
Can be merged after approval

**Comment by kremalicious** [comment]
@ssallam could you re-review please?

**Comment by kremalicious** [comment]
hey, we should still get this merged and released as actual stable release.

@ssallam I think it is all pending that you mark your change request for this PR as done

**Pull request merged**